modules/ml/src/main/java/org/apache/ignite/ml/tree/randomforest/data/impurity/GiniHistogram.java

```diff
@@ -0,0 +1,205 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.tree.randomforest.data.impurity;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.apache.ignite.ml.dataset.feature.BucketMeta;
+import org.apache.ignite.ml.dataset.feature.ObjectHistogram;
+import org.apache.ignite.ml.dataset.impl.bootstrapping.BootstrappedVector;
+import org.apache.ignite.ml.tree.randomforest.data.NodeSplit;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
+
+/**
+ * Class contains implementation of splitting point finding algorithm based on Gini metric (see https://en.wikipedia.org/wiki/Gini_coefficient)
+ * and represents * a set of histograms in according to this metric.
+ */
+public class GiniHistogram implements ImpurityComputer<BootstrappedVector, GiniHistogram> {
+    /** Bucket meta. */
+    private final BucketMeta bucketMeta;
+
+    /** Feature id. */
+    private final int featureId;
+
+    /** Sample id. */
+    private final int sampleId;
+
+    /** Hists of counters for each labels. */
+    private final ArrayList<ObjectHistogram<BootstrappedVector>> hists;
+
+    /** Label mapping to internal representation. */
+    private final Map<Double, Integer> lblMapping;
+
+    /** Bucket ids. */
+    private final Set<Integer> bucketIds;
+
+    /**
+     * Creates an instance of GiniHistogram.
+     *
+     * @param sampleId Sample id.
+     * @param lblMapping Label mapping.
+     * @param bucketMeta Bucket meta.
+     */
+    public GiniHistogram(int sampleId, Map<Double, Integer> lblMapping, BucketMeta bucketMeta) {
+        this.hists = new ArrayList<>(lblMapping.size());
+        this.featureId = bucketMeta.getFeatureMeta().getFeatureId();
+        this.sampleId = sampleId;
+        this.bucketMeta = bucketMeta;
+        this.lblMapping = lblMapping;
+
+        for (int i = 0; i < lblMapping.size(); i++)
+            hists.add(new ObjectHistogram<>(this::bucketMap, this::counterMap));
+
+        this.bucketIds = new TreeSet<>();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addElement(BootstrappedVector vector) {
+        Integer lblId = lblMapping.get(vector.getLabel());
+        hists.get(lblId).addElement(vector);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Optional<Double> get(Integer bucket) {
+        throw new NotImplementedException();
```

Comment:
What's wrong with this method?

Reply:
nothing, but it useless for RF
I could implement in this method impurity computing and use it in findBestSplit, but it is not efficient way